### ISSUE TITLE
dalvikvm: Abort

### ISSUE THREAD

**Issue by hanyman7** [termux/termux-packages#18557]
A recent software  was installed on my Samsung SM-A546U.
I can compile java code using ecj as before.
Additionally, dx --dex output=myclasses.dex
Then either add classes.dex to CLASSPATH or using -cp myclasses.dex Abort(s)

dalvikvm -cp myclasses.dex com.me.HelloWorld
Aborted

added echo statement to dalvikvm

exec /apex/com.android.art/bin/dalvikvm -Xusejit:true -Xnoimage-dex2oat -Djava.io.tmpdir=/data/data/com.termux/files/usr/tmp -cp classes.dex com.me.HelloWorld
Aborted

Interesting that both ecj and dx use same dalvikvm, but difference is that -Xcompiler mode seems to work ??? Help

Please guide me to collect any info requested
![Screenshot_20231122_074141_Settings](https://github.com/termux/termux-tools/assets/57886948/92b39bb9-263a-4bd0-8416-8e11d0f70ae4)

[termux-info.txt](https://github.com/termux/termux-tools/files/13440143/termux-info.txt)

**Comment by hanyman7** [comment]
Twaik can't thank you enough for guiding me through this. It worked
👍👍👍

**Comment by twaik** [comment]
We should patch ecj to do this automatically.

**Comment by hanyman7** [comment]
Ok, I experimented a bit.
If CLASSPATH and/or the -cp argument point to **any** files that are not write protected, dalvikvm will Abort.
So any .dex or .jar files in -cp or CLASSPATH will do. When a jar/dex file is created the individual write attributes of the elements in the jar/dex file don't matter. The only thing that matters is that the output file or jar and dex as seen by dalvikvm must be read only.

**Comment by hanyman7** [comment]
> We should patch ecj to do this automatically.

See above, I ran a few tests as I was modifying my build process and I had to ensure that all jar files (which contain .dex files) and individual.dex files that were in -cp argument or CLASSPATH as een by dalvikvm MUST be READ-ONLY.

So I am not sure that ecj is culprit.

dx --dex which used to transcode .class to dex or '.jar files contained .class files' to a dex-output is culprit. This is because it's the only way to prepare an object that is usable by dalvikvm.

One could also include multiple.dex files in a jar and use that in -cp/CLASSPATH but also the jar utility must ensure that the produced output is READ-ONLY. So jar utilities are culprit 

I hope this helps.... ecj is only converting .java to .class which are unusable as is.

**Comment by twaik** [comment]
> Ok, I experimented a bit.

It was no need to do that. #16255

**Comment by hanyman7** [comment]
> > Ok, I experimented a bit.
> 
> It was no need to do that. #16255

They are talking about am utility.
In general many users, myself included, have a lot of java code in Termux as well as Java servers. I don't use APKs

So I needed to modify my own java build environment which is a mix of jar and dex files

**Comment by twaik** [comment]
Quote:
> Android 14 requires that all dynamically loaded dex files are read-only to prevent security issues.

**Comment by hanyman7** [comment]
> Quote:
> 
> > Android 14 requires that all dynamically loaded dex files are read-only to prevent security issues.

I agree 100%, but the question is what produces the artifacts that are in -cp or CLASSPATH.
The answer is jar and dx -- dex output=...

**Comment by twaik** [comment]
Ok, now I see that we download .jar and do not build it from source. Maybe we should make it run with openjdk-17? @licy183 @truboxl ?

**Comment by twaik** [comment]
That is the problem: https://cs.android.com/android/platform/superproject/+/dd6fde63193fda968440fc0a4cd9ccee5d282ab0:art/runtime/native/dalvik_system_DexFile.cc;l=381

Android checks if current uid is `0` (root), `1000` (system) or `2000` (shell) AND if apk has write access.
Possible solution: make a little library which overrides regular `access` function which will always report no write access or override regular `getuid` which will always report `2000` (shell) and put this library to LD_PRELOAD in ecj starting script.

**Comment by twaik** [comment]
@licy183 @truboxl should we create a wrapper script for dalvikvm for overriding W^X behaviour or we should just close the issue?

**Comment by truboxl** [comment]
https://github.com/termux/termux-tools/blob/57c4adf65069cf17567330d7c00adc5b98e15419/scripts/dalvikvm.in#L13

**Comment by twaik** [comment]
Oh, so we should just create a library for that and add it to LD_PRELOAD, right?
@sylirre @Grimler91 @2096779623 pinging you because you commited to the repo.
